index.js

```diff
@@ -1,85 +1,108 @@
+const hideElement = element => {
+	element.style.visibility = 'hidden';
+};
+
+const getBoundingClientRect = element => {
+	const {height, width, x, y} = element.getBoundingClientRect();
+	return {height, width, x, y};
+};
+
+module.exports = async (url, opts) => {
 	opts = Object.assign({
-		delay: 0,
-		scale: 1,
-		format: 'png'
+		cookies: [],
+		fullPage: true,
+		hide: [],
+		width: 1920,
+		height: 1080
 	}, opts);
 
-	const args = Object.assign(opts, {
-		url,
-		width: size.split(/x/i)[0] * opts.scale,
-		height: size.split(/x/i)[1] * opts.scale,
-		cookies: handleCookies(opts.cookies, url),
-		format: opts.format === 'jpg' ? 'jpeg' : opts.format,
-		css: /\.css$/.test(opts.css) ? fs.readFileSync(opts.css, 'utf8') : opts.css,
-		script: /\.js$/.test(opts.script) ? fs.readFileSync(opts.script, 'utf8') : opts.script
-	});
-
-	const cp = phantomBridge(path.join(__dirname, 'stream.js'), [
-		'--ignore-ssl-errors=true',
-		'--local-to-remote-url-access=true',
-		'--ssl-protocol=any',
-		JSON.stringify(args)
-	]);
-
-	const stream = base64Stream.decode();
-
-	process.stderr.setMaxListeners(0);
-
-	cp.stderr.setEncoding('utf8');
-	cp.stdout.pipe(stream);
-
-	byline(cp.stderr).on('data', data => {
-		data = data.trim();
-
-		if (/ phantomjs\[/.test(data)) {
-			return;
-		}
-
-		if (/http:\/\/requirejs.org\/docs\/errors.html#mismatch/.test(data)) {
-			return;
-		}
-
-		if (data.startsWith('WARN: ')) {
-			stream.emit('warning', data.replace(/^WARN: /, ''));
-			stream.emit('warn', data.replace(/^WARN: /, '')); // TODO: deprecate this event in v5
-			return;
-		}
-
-		if (data.length > 0) {
-			const err = new Error(data);
-			err.noStack = true;
-			cp.stdout.unpipe(stream);
-			stream.emit('error', err);
-		}
-	});
-
-	return stream;
+	const uri = isUrl(url) ? url : fileUrl(url);
+	const {
+		cookies, crop, format, headers, height, hide, keepAlive, password, scale,
+		script, selector, timeout, transparent, userAgent, username, width
+	} = opts;
+
+	opts.type = format === 'jpg' ? 'jpeg' : format;
+
+	if (crop) {
+		opts.fullPage = false;
+	}
+
+	if (timeout) {
+		opts.timeout = timeout * 1000;
+	}
+
+	if (transparent) {
+		opts.omitBackground = true;
+	}
+
+	const browser = opts.browser || await puppeteer.launch();
+	const page = await browser.newPage();
+	const viewport = {
+		height,
+		width,
+		deviceScaleFactor: typeof scale === 'number' ? scale : null
+	};
+
+	if (username && password) {
+		await page.authenticate({username, password});
+	}
+
+	if (Array.isArray(cookies) && cookies.length > 0) {
+		await Promise.all(cookies.map(x => page.setCookie(parseCookie(x))));
+	}
+
+	if (typeof headers === 'object') {
+		await page.setExtraHTTPHeaders(headers);
+	}
+
+	if (userAgent) {
+		await page.setUserAgent(userAgent);
+	}
+
+	await page.setViewport(viewport);
+	await page.goto(uri, opts);
+
+	if (script) {
+		const fn = isUrl(script) ? page.addScriptTag : script.endsWith('.js') ? page.injectFile : page.evaluate;
```

Comment:
I wonder if we could write this line like this

```
const fn = isUrl(script) || script.endsWith('.js') ? page.addScriptTag : page.evaluate;
```

A script tag works for a local file as well right? If not, I think we should wrap the second block in parentheses for readability.

```js
const fn = isUrl(script) ? page.addScriptTag : (script.endsWith('.js') ? page.injectFile : page.evaluate);
```